### ISSUE TITLE
Feature request: adding TA's to teams without having to resort to owners

### ISSUE THREAD

**Issue by LudwikJaniuk** [repobee/repobee#416]

**Comment by slarse** [comment]
Generalized TA management is a good idea. We've always operated on the idea of ultimate trust in TAs, but I can see how some may not want that.

I think this can be entirely handled with a plugin. It may be core-feature worthy, but a plugin is a good start. The best solution is probably to just add TAs to the student teams, which will work fine both on GitHub and GItLab. Some expansion of the API classes would however be required for some advanced functionality. Here's a rundown.

* Adding TA's to student teams to give them access
    - Borderline trivial to do using `API.ensure_teams_and_members` and the `STUDENTS` base parser tagged on to the extension command
* Removing TA's from teams (i.e. revoking access)
    - Requires expansion of the platform API wrapper, but it's a simple expansion
* Config file integration
    - Easy to add with existing plugin system

**Comment by slarse** [comment]
@LudwikJaniuk We are implementing this over the summer in a proper plugin. Do you have any feedback on the prototype I sent to you?

**Comment by LudwikJaniuk** [comment]
A "release" version should have the following features IMO:
 * specify TA's in the config file
 * adding another TA later (people get hired whenever)
 * removing TA's (why make it hard to follow good practice) ;)
I haven't researched GitHub's access model so much, but maybe there's a cleaner way to do it btw? Like roles or having a dedicated TA team with access, but so that adding/removing could be done thru GitHub interface? I think it's good if repobee can keep as many features _outside of_ repobee as possible :)

Thanks for giving this feature attention!

**Comment by slarse** [comment]
All of the features you listed are quite easily accomplished, except for making it manageable both via RepoBee and the web UI.

> I haven't researched GitHub's access model so much, but maybe there's a cleaner way to do it btw? Like roles or having a dedicated TA team with access, adding/removing could be done thru GitHub interface? I think it's good if repobee can keep as many features outside of repobee as possible

It's possible to create a TA team that gets access to all existing repositories. Then, adding TAs to it could be done in the web interface as well as with RepoBee. It's a bit harder to automatically update access when new repos are created, as RepoBee must be aware of there potentially being a TA team and add new repositories to it. It could potentially be done if we add a new hook to the `setup` command.

We probably need to make sure that it works with GitLab as well, as RepoBee supports that as well. It's however possible that we don't even need a plugin like this for GitLab, as the access control in GitLab is much better designed than it is in GitHub, so then we could make a tailored GitHub plugin.

We'll see what we come up with, at the very list we'll implement your three bullet points.

**Comment by LudwikJaniuk** [comment]
Good to hear!
Best regards,
//Ludvig Janiuk


Den mån 15 juni 2020 kl 10:02 skrev Simon Larsén <notifications@github.com>:

**Comment by slarse** [comment]
@LudwikJaniuk So, this took a while, but I finally got around to implementing it. It's pretty close to your initial request. See #674 

I'm not sure if this is even relevant to your line of work anymore, but it was a nice feature request and it will be used in other courses :)
